src/elements/EditableTable.tsx

```diff
@@ -0,0 +1,342 @@
+            } else {
+                total[col.name] = col.columnCalculation(values);
+            }
+        }
+    }
+    return total;
+};
+
+const calculateSum = (values: any[]): number => {
+    const numericVals = values.filter((x) => typeof x == "number");
+    return numericVals.length > 0 ? numericVals.reduce((s: number, a: number) => s + a, 0) : 0;
+};
+
+const calculateAverage = (values: any[]): number => {
+    const numericVals = values.filter((x) => typeof x == "number");
+    const sum = numericVals.length > 0 ? numericVals.reduce((s: number, a: number) => s + a, 0) : 0;
+    return sum / numericVals.length || 0;
+};
+
+const formatValue = (value: any, column: Column): any => {
+    if (column.type === "number" && typeof value === "number") {
+        if (column.numberFormat) {
+            return format(column.numberFormat as string, value);
+        } else {
+            return value;
+        }
+    } else if (typeof value === "string") {
+        return value;
+    } else if (value instanceof Date) {
+        return value.toDateString();
+    } else if (Object.prototype.hasOwnProperty.call(value, "toString")) {
+        return value.toString();
+    }
+    return "-";
+};
+
+/**
+ * A table form element.
+ * @displayName Editable Table
+ * @description A table form element that allows for inline editing and tabular calculation.
+ * @param props The props that will be provided by the Workflow runtime.
+ */
+function EditableTableElement(props: EditableTableElementProps): React.ReactElement | null {
+    const {
+        rows,
+        cols,
+        maxHeight,
+        maxWidth,
+        size,
+        stickyHeader,
+        footerLabel,
+        onMouseEnter,
+        onMouseLeave,
+        raiseEvent,
+        onRowEdit,
+    } = props;
+
+    const [data, setData] = useState<RowData[]>([]);
+    const [editingRows, setEditingRows] = useState<Record<number, RowData>>({});
+
+    useEffect(() => {
+        setData(rows);
+    }, [rows]);
+
+    const includeFooter = cols.some((x) => x.columnCalculation);
+    const canEdit = cols.some((x) => x.editable);
+    const footerRow: FooterRow | undefined = includeFooter
+        ? calculateFooterRow(cols, rows)
+        : undefined;
+
+    const handleMouseEnter = (event: React.MouseEvent, row: RowData) => {
+        raiseEvent("custom", { eventType: "mouseEnter", row });
+        onMouseEnter?.(row);
+    };
+
+    const handleMouseLeave = (event: React.MouseEvent, row: RowData) => {
+        raiseEvent("custom", { eventType: "mouseLeave", row });
+        onMouseLeave?.(row);
+    };
+
+    const handleCellChange = (value: string | number, row: RowData, column: Column) => {
+        if (column.type === "number") {
+            row[column.name] = Number(value);
+        } else {
+            row[column.name] = value;
+        }
+        setEditingRows({ ...editingRows });
+    };
+
+    const handleToggleEdit = (event: React.MouseEvent, index: number, rows: RowData[]) => {
```

Comment:
Please split this into separate `handleBeginEdit` and `handleCompleteEdit` methods?

Also it looks like the `rows` parameter is not necessary because we were just passing in the `rows` state which we can reference directly here.

Reply:
Done